File: android/src/main/location-engine-google/org/maplibre/reactnative/location/engine/GoogleLocationEngineProvider.java
package org.maplibre.reactnative.location.engine;

import android.content.Context;
import android.util.Log;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;

import org.maplibre.android.location.engine.LocationEngine;
import org.maplibre.android.location.engine.LocationEngineProxy;

public class GoogleLocationEngineProvider implements LocationEngineProvidable {
    private static final String LOG_TAG = "GoogleLocationEngineProvider";

    public LocationEngine getLocationEngine(Context context) {
        if (GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context) == ConnectionResult.SUCCESS) {
            LocationEngine locationEngine = new LocationEngineProxy<>(new GoogleLocationEngineImpl(context.getApplicationContext()));
            Log.d(LOG_TAG, "GoogleLocationEngine will be used.");
            return locationEngine;
        } else {
            return new DefaultLocationEngineProvider().getLocationEngine(context);
        }
    }
}

File: package/android/src/main/location-engine-google/org/maplibre/reactnative/location/engine/GoogleLocationEngineProvider.kt
package org.maplibre.reactnative.location.engine

import android.content.Context
import android.util.Log
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.location.LocationCallback
import org.maplibre.android.location.engine.LocationEngine
import org.maplibre.android.location.engine.LocationEngineProxy

class GoogleLocationEngineProvider : LocationEngineProvidable {
    override fun getLocationEngine(context: Context): LocationEngine {
        if (GoogleApiAvailability
                .getInstance()
                .isGooglePlayServicesAvailable(context) == ConnectionResult.SUCCESS
        ) {
            val locationEngine: LocationEngine =
                LocationEngineProxy<LocationCallback?>(GoogleLocationEngineImpl(context.applicationContext))
            Log.d(LOG_TAG, "GoogleLocationEngine will be used.")
            return locationEngine
        } else {
            return DefaultLocationEngineProvider().getLocationEngine(context)
        }
    }

    companion object {
        private const val LOG_TAG = "GoogleLocationEngineProvider"
    }
}
